module/VuFind/src/VuFind/Search/Solr/Explanation.php

```diff
@@ -0,0 +1,531 @@
+                $percentage = 100 * $currentValue / $this->baseScore;
+            } else {
+                $percentage = 0;
+            }
+
+            // get fieldModifier and remove unused higher level lines
+            $fieldModifier = null;
+            if (str_contains($description, 'const weight')) {
+                $fieldModifier = 0;
+            }
+            while (!empty($lines) && $curLevel < $this->getLevel(end($lines))) {
+                $childLine = array_pop($lines);
+                $childInfo = $this->parseLine($childLine);
+                $childValue = $childInfo['value'];
+                $childDescription = $childInfo['description'];
+                if ($childDescription === ' boost') {
+                    $fieldModifier = $childValue;
+                }
+            }
+
+            // add to rest if lower than min percentage
+            $explainElement = $this->parseExplainElement($currentValue, $description, $percentage, $fieldModifier);
+            if ($percentage < $this->getMinPercentage()) {
+                $this->explanationForRest[] = $explainElement;
+            } else {
+                $this->explanation[] = $explainElement;
+            }
+        }
+        return $lines;
+    }
+
+    /**
+     * Returns indent of a line.
+     *
+     * @param string $line Line
+     *
+     * @return int
+     */
+    protected function getLevel($line)
+    {
+        return (strlen($line) - strlen(ltrim($line))) / 2;
+    }
+
+    /**
+     * Gets all lines with one level higher than the parent line.
+     *
+     * @param array $lines Lines
+     * @param int   $level Level
+     *
+     * @return array
+     */
+    protected function getChildLines($lines, $level)
+    {
+        $res = [];
+        while (!empty($lines) && $this->getLevel(end($lines)) > $level) {
+            $line = array_pop($lines);
+            if ($this->getLevel($line) == $level + 1) {
+                $res[] = $line;
+            }
+        }
+        return $res;
+    }
+
+    /**
+     * Extracts value and description of a line.
+     *
+     * @param string $line Line
+     *
+     * @return array
+     */
+    protected function parseLine($line)
+    {
+        $info = explode('=', $line, 2);
+        return [
+            'value' => floatval($info[0]),
+            'description' => $info[1],
+        ];
+    }
+
+    /**
+     * Unites all infos of a match to an explainElement.
+     *
+     * @param float  $value         Value
+     * @param string $description   Description
+     * @param float  $percentage    Percentage
+     * @param float  $fieldModifier Field Modifier
+     *
+     * @return array
+     */
+    protected function parseExplainElement($value, $description, $percentage, $fieldModifier)
+    {
+        $res = [
+            'value' => $value,
+            'percent' => $percentage,
+        ];
+        $pattern = '/weight\((?<fieldName>[^:]+):(?<fieldValue>\"([^\W]+\s?)+[^\\W]+\"|[^\W]+)(.+?(?= in))?/';
```

Comment:
Should match in UTF-8 mode `/.../u`. Otherwise this messes up non-ascii characters. This should be `kylä`:
![kuva](https://github.com/vufind-org/vufind/assets/1327659/26f57956-fc8e-43b7-922c-8963575d9cef)